service/service.go

```diff
@@ -686,12 +687,65 @@ func (s *Service) MakeNewSecret(ctx context.Context, req *pbsvc.UserRequest) (*p
 }
 
 // VerifyEmailToken checks if received token is found in the email_tokens table.
-// If found and token is NOT expired, returns OK.
-// If found, but token IS expired, the token row with expired token will be deleted,
-// service will generate a new email token and insert into email_tokens table,
-// and return ERROR with token expired message.
-// If token is not found, return ERROR with token does not exist message.
+// If found and token is NOT expired, deletes token row and returns OK.
+// If found, but token IS expired, it will return a expired token error.
+// Additionally for expired tokens, if user is new, it will delete token AND user row, else just deletes the token row.
+// If token is not found, return error with token does not exist message.
 func (s *Service) VerifyEmailToken(ctx context.Context, req *pbsvc.UserRequest) (*pbsvc.UserResponse, error) {
-	// TODO
-	return nil, nil
+	logger.RequestService("VerifyEmailToken")
+
+	if ok := serviceStateLocker.isStateAvailable(); !ok {
+		logger.Error(consts.VerifyEmailToken, consts.ErrServiceUnavailable.Error())
+		return nil, consts.ErrStatusServiceUnavailable
+	}
+
+	if req == nil {
+		return nil, consts.ErrStatusNilRequestUser
+	}
+
+	if err := refreshDBConnection(); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	emailToken := req.GetIdentification().GetToken()
+	if emailToken == "" {
+		return nil, status.Error(codes.InvalidArgument, authconst.ErrEmptyToken.Error())
+	}
+
+	// find matching email token row
+	retrievedToken, err := getEmailTokenRow(emailToken)
+	if err != nil {
+		return nil, err
+	}
+
+	// delete token row
+	if err := deleteEmailTokenRow(retrievedToken.uuid); err != nil {
+		logger.Error(consts.VerifyEmailToken, consts.MsgErrDeletingEmailToken)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// if token is expired
+	if time.Now().Unix() >= retrievedToken.expirationTimestamp {
```

Comment:
What if time is 0 or less?

Reply:
when you generate expirationTimestamp using `generateExpirationTimestamp`, it prevents you from adding time anything below or equal to 0.

So assuming developer always generates expirationTimestamp using the function below, I don't think we will have problem?

```
func generateExpirationTimestamp(currentTimestamp time.Time, addDays int) (*time.Time, error) {
	if currentTimestamp.IsZero() {
		return nil, consts.ErrInvalidTimeStamp
	}

	if addDays <= 0 {
		return nil, consts.ErrInvalidNumberOfDays
	}
```

Reply:
in this case, there is high coupling which is not ideal. additionally, relying that `generateExpirationTimestamp` will not be modified and reviewed by different developers is bad assumption. Function has to be cohesive, meaning it is good to what is supposed to be doing and not relying on external implementation.

Reply:
> So **assuming** developer always generates expirationTimestamp using the function below, I don't think we will have problem?

That can be a problem